lib/shared/src/encode_key_value.rs

```diff
@@ -0,0 +1,729 @@
+        value: &T,
+    ) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        value.serialize(self.child(key))
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        Ok(())
+    }
+}
+
+struct IndexedKeyValueSerializer<'a> {
+    index: usize,
+    ser: KeyValueSerializer<'a>,
+}
+
+impl<'a> IndexedKeyValueSerializer<'a> {
+    fn process<T: ?Sized + Serialize>(&mut self, data: &T) -> Result<(), EncodingError> {
+        let key = self.index;
+        self.index += 1;
+        data.serialize(self.ser.child(key))
+    }
+}
+
+impl<'a> SerializeTuple for IndexedKeyValueSerializer<'a> {
+    type Ok = ();
+    type Error = EncodingError;
+
+    fn serialize_element<T: ?Sized>(&mut self, value: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        self.process(value)
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        Ok(())
+    }
+}
+
+impl<'a> SerializeSeq for IndexedKeyValueSerializer<'a> {
+    type Ok = ();
+    type Error = EncodingError;
+
+    fn serialize_element<T: ?Sized>(&mut self, value: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        self.process(value)
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        Ok(())
+    }
+}
+
+impl<'a> SerializeTupleStruct for IndexedKeyValueSerializer<'a> {
+    type Ok = ();
+    type Error = EncodingError;
+
+    fn serialize_field<T: ?Sized>(&mut self, value: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        self.process(value)
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        Ok(())
+    }
+}
+
+impl<'a> SerializeTupleVariant for IndexedKeyValueSerializer<'a> {
+    type Ok = ();
+    type Error = EncodingError;
+
+    fn serialize_field<T: ?Sized>(&mut self, value: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        self.process(value)
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        Ok(())
+    }
+}
+
+struct KeyedKeyValueSerializer<'a> {
+    key: Option<String>,
+    ser: KeyValueSerializer<'a>,
+}
+
+impl<'a> SerializeMap for KeyedKeyValueSerializer<'a> {
+    type Ok = ();
+    type Error = EncodingError;
+    fn serialize_key<T: ?Sized>(&mut self, key: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        use serde_json::{to_value, Value};
+        match to_value(key) {
+            Ok(Value::String(key)) => {
+                self.key = Some(key);
+                Ok(())
+            }
+            _ => Err(EncodingError::KeyNotString),
+        }
+    }
+
+    fn serialize_value<T: ?Sized>(&mut self, value: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        let key = self.key.take().expect("Key must be present.");
+        value.serialize(self.ser.child(key))
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
It appears all tests are `unwrap`ing the `to_string` call. In what situation can this call fail, and can we adda a test for that? If there are none, should we `expect` inside `to_string` itself and not have it return a `Result`?

I'm asking because I noticed we're adding `expect` to all the non-test calls as well.

Reply:
It depends on the `V: Serialized` passed to `to_string`, if it contains non `String` keys then it can fail. Vrl and Vector `Value` do always have `String` as keys so we `expect`. But true, a test for that and a comment would be nice.